### PR TITLE
Add more crypto keys

### PR DESCRIPTION
**Pull request by LunaMoo**
Found in https://www.consolex.ru/forum/showthread.php?t=85431
Probably doesn't fix anything :/
includes keys needed for #12343, personally tried only  0x2E5E10F0, key_2E5E10F0, 0x48, 5, xor_2E5E10F0 which is used for game update I had and the result is just slightly less wrong than currently - file size is correct, but still no proper elf, not sure if it's wrong key or the code that uses it.
 And yeah the xor is repeated, took pauth xor as an example, where it also is repeated and just kept copies.

 Didn't bother to look at other keys on that forum as generally PPSSPP can decrypt everything except game updates already.

 BTW some those xor keys are listed instead as (OPNSSMP.BIN), soo maybe xor there is just wrong or matches only somebody's game and proper xor key needs to be somehow taken from that file(it's on game disc). Edit: thinking about it, probably not true, since we can use homebrew to decrypt at least 0x2E5E10F0 on real PSP without ever placing the game on psp.

 Changing to draft since it's kind of pointless to merge as it doesn't really improve situation. Might help someone in the future.

### CODE REVIEW

**Comment by hrydgard** [comment]
Hm, this might still be worth merging if it gets us correct file sizes, should get rid of this crash I'm seeing in my reports:

```
#00  pc 0x0000000000089cd0  /apex/com.android.runtime/lib64/bionic/libc.so (abort+164)
#01  pc 0x0000000000fe482c  arm64/libppsspp_jni.so (abort_message) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#02  pc 0x0000000000fe4984  arm64/libppsspp_jni.so (demangling_terminate_handler()) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#03  pc 0x0000000000fe1874  arm64/libppsspp_jni.so (std::__terminate(void (*)())) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#04  pc 0x0000000000fe0ea0  arm64/libppsspp_jni.so (__cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*)) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#05  pc 0x0000000000fe0dfc  arm64/libppsspp_jni.so (__cxa_throw+112) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#06  pc 0x0000000000fd75bc  arm64/libppsspp_jni.so (operator new(unsigned long)+96) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#07  pc 0x000000000052cb28  arm64/libppsspp_jni.so (NPDRMDemoBlockDevice::NPDRMDemoBlockDevice(FileLoader*)+352) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#08  pc 0x000000000052bb48  arm64/libppsspp_jni.so (constructBlockDevice(FileLoader*)+412) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#09  pc 0x00000000008a5ca4  arm64/libppsspp_jni.so (GameInfoWorkItem::Run()+644) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#10  pc 0x000000000085bfa4  arm64/libppsspp_jni.so (WorkerThreadFunc(GlobalThreadContext*, TaskThreadContext*)) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#11  pc 0x000000000085dab4  arm64/libppsspp_jni.so (void* std::__ndk1::__thread_proxy<std::__ndk1::tuple<std::__ndk1::unique_ptr<std::__ndk1::__thread_struct, std::__ndk1::default_delete<std::__ndk1::__thread_struct>>, void (*)(GlobalThreadContext*, TaskThreadContext*), GlobalThreadContext*, TaskThreadContext*>>(void*)+48) (BuildId: 7918a182fa2ad427379f7705c2f0bcb662ac20f4)
#12  pc 0x00000000000eb7a8  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+64)
```

**Comment by LunaMoo** [comment]
If proper file size is enough to solve some crash without introducing new ones then go for it, but I was thinking about just closing it and wait until the Kirk info becomes public and have a proper fix ~ https://github.com/hrydgard/ppsspp/issues/12344#issuecomment-1877898070

**Comment by hrydgard** [comment]
It can't really hurt either, so let's get it in.

**Comment by Linblow** [comment]
`key_2E5E10F0` is correct and should get you the correct decrypted ELF.

As for `key_2E5E11F0`, `key_2E5E80F0` and `key_2E5E90F0` they are a bit special.
They are part of the resmgr tags/keys which the game can dynamically set via the `OPNSSMP.BIN` file, if any.
When this file is not filled with zeroes, it contains a small encrypted executable (and KL4E compressed once decrypted).
In that case, VSH will decrypt & decompress it, and run it at game boot before `EBOOT.BIN` gets run.

This small `OPNSSMP.BIN` program contains and sets the tags/keys for:
1. the game's `EBOOT.BIN` executable
2. the game's `PBOOT.PBP` DATA.PSP encrypted patch. 

It's done via a call to this function (mesg_led API):

```c
/**
 * Set tag/seed for EBOOT/PBOOT module decryption.
 * Imported in user mode by OPNSSMP.BIN module for some games.
 * Exported twice, for user and kernel mode. 
 * @param opn EBOOT/PBOOT keys (size: 0x78)
 * @return 0 success; < 0 internal mseg_led error
 */
s32 sceResmgr_8E6C62C8(const SceResmgrOpnssmp *opn); 
```
The data structure is as follows:
```c
/**
 * ~PSP module tag with a custom seed key.
 */
struct SceResmgrModuleTag
{
   /** ~PSP module tag. */
   ScePSPModuleTag tag;
   /** Custom 16-bytes seed key. */
   ScePSPModuleKey key;
}; // size: 0x14

/**
 * Set of encrypted keys for decrypting an encrypted game boot executable.
 * 
 * The plain (M) and ciphertext (C) data is constructed as follows:
 * M = KIRK_7( C XOR resmgrKey2 ) XOR resmgrKey2
 * C = KIRK_4( M XOR resmgrKey2 ) XOR resmgrKey2
 * It uses KIRK key index 0x56.
 * This structure's data will be internally decrypted in sceResmgr_8E6C62C8().
 */ 
struct SceResmgrBootKeys
{
   /** EBOOT game executable tag/key. */
   SceResmgrModuleTag eboot;
   /** Patched-BOOT game executable tag/key. */
   SceResmgrModuleTag pboot;
   /** Necessary padding since the whole structure is AES encrypted. */
   u8 padding[8];
}; // size: 0x30
```
As you can see, the structure is rather trivial, and can be easily decrypted.
These are the correct *default* resmgr tags/keys (6.61 02g):
```c
/* The 6 resmgr tags *can* be changed at runtime prior to game boot via OPNSSMP.BIN */
// default tag: D91611F0
static u8 resmgr_eboot1_key[]  = {0x61,0xb0,0xc0,0x58,0x71,0x57,0xd9,0xfa,0x74,0x67,0x0e,0x5c,0x7e,0x6e,0x95,0xb9};
// default tag: D91680F0
static u8 resmgr_eboot2_key[]  = {0x2c,0x22,0x9b,0x12,0x36,0x74,0x11,0x67,0x49,0xd1,0xd1,0x88,0x92,0xf6,0xa1,0xd8};
// default tag: D91690F0
static u8 resmgr_eboot3_key[]  = {0x42,0x61,0xe2,0x57,0x94,0x49,0x42,0xb5,0xaa,0x6d,0x0d,0x08,0x3d,0x24,0xf7,0x4b};
// default tag: 2E5E11F0
static u8 resmgr_pboot1_key[]  = {0x75,0xeb,0xe8,0x43,0xf4,0x87,0x8f,0xd0,0x14,0x7f,0x7e,0x39,0xad,0xaf,0x04,0x9d};
// default tag: 2E5E80F0
static u8 resmgr_pboot2_key[]  = {0x0f,0x74,0xaf,0x43,0x75,0xcd,0xda,0x39,0x81,0x56,0xd9,0x61,0x3e,0x16,0xc8,0x92};
// default tag: 2E5E90F0
static u8 resmgr_pboot3_key[]  = {0x67,0xe4,0x8f,0x4c,0x08,0xa0,0x7d,0xb1,0x5f,0x51,0xa7,0x72,0x98,0xa8,0x2d,0x7e};
/* Resmgr PBOOT tags have a constant secondary key though. */
static const u8 resmgr_pboot_key2[] = {0x69,0xba,0x55,0x34,0xf0,0xc0,0xd6,0x71,0xe3,0x1f,0xdb,0x97,0xe0,0x7c,0xd2,0x2a};

/* The primary key AND tag CAN be changed when OPNSSMP is run, the second key never changes. */
static TAG_INFO2 g_resmgrTagInfo[] =
{
	{ 0x2E5E11F0, resmgr_pboot1_key, 0x48, 5,  resmgr_pboot_key2 },
	{ 0x2E5E80F0, resmgr_pboot2_key, 0x48, 7,  resmgr_pboot_key2 },
	{ 0x2E5E90F0, resmgr_pboot3_key, 0x48, 10, resmgr_pboot_key2 },
	{ 0xD91611F0, resmgr_eboot1_key, 0x5d, 2 },
	{ 0xD91680F0, resmgr_eboot2_key, 0x5d, 6 },
	{ 0xD91690F0, resmgr_eboot3_key, 0x5d, 9 },
};
```
Note this was obtained from the reversed `mesg_led_02g` kernel module.
Basically, the game will overwrite one pboot tag/key, and one eboot tag/key.
I'm working on it, as part of the game patch implementation PR. Stay tuned.

**Comment by hrydgard** [comment]
Very cool, thanks for the info Linblow!
